### PR TITLE
Sia Integration

### PR DESCRIPTION
**Pull request by dvstate**
## Description
This code change adds a Sia gateway to Minio server, allowing data to be stored on the Sia network using Minio as a frontend. It changes nothing functional with the existing Minio codebase. It simply adds an additional gateway to allow users to store content on Sia.

Documentation for using the Sia gateway, which can be incorporated into Minio's documentation, can be found here:
https://github.com/dvstate/minio/blob/master/README.md

## Motivation and Context
To allow Minio users to utilize Sia for inexpensive backend storage.

## How Has This Been Tested?
New methods were individually tested as they were written.
Basic Go tests have been written for some of the Sia functionality.
Live servers using the Sia gateway have been operational for a couple weeks now, and all reported bugs, issues, and annoyances have been addressed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/gateway-sia-anonymous.go
Comment (harshavardhana) [inline cmd/gateway-sia-anonymous.go]: Missing license header..

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: License header..

File: cmd/gateway-sia-helpers.go
Comment (harshavardhana) [inline cmd/gateway-sia-helpers.go]: License.. header.

File: cmd/server-main.go
Comment (harshavardhana) [inline cmd/server-main.go]: We should use globalMinioPort here

File: cmd/gateway-sia_test.go
Comment (harshavardhana) [inline cmd/gateway-sia_test.go]: License header..

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: Delete bucket shouldn't delete all top level objects.. only multipart i.e temporary objects can be deleted. Otherwise DeleteBucket() should return ENOTEMPTY.  Look at the fs-v1.go implementation for more inputs.

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: Lets actually explore on removing this dependency and moving to either https://github.com/pingcap/tidb or https://github.com/boltdb/bolt 

BoltDB is more native in its feel APIs are sort of synonymous and its quite small. 

https://github.com/pingcap/tidb is more feature rich and SQL support. 

I am leaning towards BoltDB but its your choice. We really need to get rid of CGO dependency.

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Initial comments.. Thanks for sending the PR @dvstate

File: cmd/gateway-sia-helpers.go
Comment (vadmeste) [inline cmd/gateway-sia-helpers.go]: I am not sure why we use abs() here instead of using directly filepath.Abs()

File: cmd/gateway-sia.go
Comment (vadmeste) [inline cmd/gateway-sia.go]: Do you mean `host` here instead of `127.0.0.1:9980` ?

**Comment by vadmeste** [comment]
@dvstate, can you briefly explain why we are using FS layer and sql in this PR? I mean a little story of the design.

**Comment by aleksclark** [comment]
@vadmeste the FS layer is used because fetching a file from the Sia network is very slow, so @dvstate has implemented a cache to hold frequently accessed files. Sia doesn't really have the concept of buckets either, thus using sql to store the map of objects -> buckets, as well as the object's status on the Sia network.

**Comment by tlightsky** [comment]
@vadmeste https://github.com/dvstate/minio/blob/master/README.md#about-the-sia-cache this might help understand

File: cmd/gateway-sia-anonymous.go
Comment (dvstate) [reply cmd/gateway-sia-anonymous.go]: Included in latest commit. If license needs changed, just let me know.

File: cmd/gateway-sia-cachelayer.go
Comment (dvstate) [reply cmd/gateway-sia-cachelayer.go]: Included in latest commit. If license needs changed, just let me know.

File: cmd/gateway-sia-cachelayer.go
Comment (dvstate) [reply cmd/gateway-sia-cachelayer.go]: The latest commit uses BoltDB and the CGO dependency has been eliminated.

File: cmd/gateway-sia-cachelayer.go
Comment (dvstate) [reply cmd/gateway-sia-cachelayer.go]: DeleteBucket should now behave as you described.

File: cmd/gateway-sia-helpers.go
Comment (dvstate) [reply cmd/gateway-sia-helpers.go]: Included in latest commit. If license needs changed, just let me know.

File: cmd/server-main.go
Comment (dvstate) [reply cmd/server-main.go]: Change made in latest commit.

File: cmd/gateway-sia_test.go
Comment (dvstate) [reply cmd/gateway-sia_test.go]: Included in latest commit. If license needs changed, just let me know.

File: cmd/gateway-sia-helpers.go
Comment (dvstate) [reply cmd/gateway-sia-helpers.go]: Changes made as requested in latest commit.

File: cmd/gateway-sia.go
Comment (dvstate) [reply cmd/gateway-sia.go]: Change made as requested so that the Sia daemon api address can be specified on command line.

File: browser/app/js/actions.js
Comment (harshavardhana) [inline browser/app/js/actions.js:496]: what is this change for? looks like unrelated. Would it be possible to send it later after this PR is integrated into master? . Perhaps you can send this separately as well? @dvstate

File: cmd/gateway-main.go
Comment (harshavardhana) [inline cmd/gateway-main.go]: Fix the GCS should say Sia backend here.. and perhaps also should list out an example setup here.

File: cmd/gateway-handlers.go
Comment (harshavardhana) [inline cmd/gateway-handlers.go:531]: I would suggest you to remove the bucket policy related changes we decided to do it differently and not keep/manage any in memory state for gateway code.  So you can perhaps just remove my PR in this regard.  We can say currently bucket policies do not work on Sia backend and fix it later.

File: cmd/gateway-sia-anonymous.go
Comment (harshavardhana) [reply cmd/gateway-sia-anonymous.go]: Yes @dvstate it needs to say `Minio Cloud Storage` and not `David Gore`.  Will check back with @abperiasamy on this for final call.

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: Move the `github.com` imports as mentioned here - you can also use `goimports` to do this for you. https://github.com/golang/go/wiki/CodeReviewComments#imports

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: ~~Does this insert the entire object to `boltDB` ? if yes have you tested with large files? - Would it be possible to use boltDB just for namespace and use regular filesystem for staging? it would be faster..~~

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [inline cmd/gateway-sia-cachelayer.go]: Should this be just `MakeBucket` ? why a new API?

File: cmd/gateway-sia-db.go
Comment (harshavardhana) [inline cmd/gateway-sia-db.go]: Looks like some code formatting issues here..

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Some review comments.. testing right now.

**Comment by harshavardhana** [comment]
Also you should run `gofmt` on your code. 

```
$ gofmt -s -l cmd/
cmd/gateway-sia-anonymous.go
cmd/gateway-sia-cachelayer.go
cmd/gateway-sia-db.go
cmd/gateway-sia-errors.go
cmd/gateway-sia-helpers.go
cmd/gateway-sia-unsupported.go
cmd/gateway-sia.go
cmd/gateway-sia_test.go
```

File: browser/app/js/actions.js
Comment (dvstate) [reply browser/app/js/actions.js:496]: The section of code added to actions.js is to allow the javascript alert dialog to display a descriptive and meaningful error message. Otherwise the user doesn't know specifically what the problem is. Quite a few users requested this ability during testing.

File: cmd/gateway-handlers.go
Comment (dvstate) [reply cmd/gateway-handlers.go:531]: So you want me to remove all the changes that you made to support bucket policies? There are multiple servers running now by Sia users that depend on bucket policy support. I guess they could continue using a branch of my version of Minio until policies are fully supported, but I do imagine you'll get quite a bit of feedback from users regarding policies not working. :)

Comment (harshavardhana) [reply cmd/gateway-handlers.go:531]: Yes the problem is not that, the problem is that its not a correct approach and has lots of holes in it.  It is okay to loose this ability while we address the overall architecture.  I reached out to you on slack, perhaps we can discuss on other things in the PR.

File: cmd/gateway-sia-cachelayer.go
Comment (harshavardhana) [reply cmd/gateway-sia-cachelayer.go]: I see that its just namespace..

File: cmd/gateway-sia-anonymous.go
Comment (dvstate) [reply cmd/gateway-sia-anonymous.go]: Fixed in most recent commit.

File: cmd/gateway-sia-helpers.go
Comment (dvstate) [reply cmd/gateway-sia-helpers.go]: Fixed in most recent commit.

File: cmd/gateway-sia_test.go
Comment (dvstate) [reply cmd/gateway-sia_test.go]: Fixed in most recent commit.

File: cmd/gateway-sia-cachelayer.go
Comment (dvstate) [reply cmd/gateway-sia-cachelayer.go]: Correct. DB is only used to store small metadata for buckets and objects.

File: cmd/gateway-main.go
Comment (dvstate) [reply cmd/gateway-main.go]: Corrected. Also added all ENV variables and descriptions and defaults to the printout message.

**Comment by mannp** [comment]
Looking forward to this one being finalised @harshavardhana :)

**Comment by mannp** [comment]
Wondered how this was progressing @harshavardhana :) 

I have a couple of issues with restic (https://github.com/dvstate/minio/issues) and I can't expect them to take a look while @dvstate 's PR is not 'official'.

Sia brought me to Minio (and I am glad it did) so I am keen to explore Minio and Sia some more.

**Comment by zachherbert** [comment]
Hi all, do we have any updates on this PR?

**Comment by harshavardhana** [comment]
> Hi all, do we have any updates on this PR?

@zachherbert this is waiting on a re-factor.

**Comment by mtlynch** [comment]
@dvstate are you still actively working on this?

**Comment by mtlynch** [comment]
>@zachherbert this is waiting on a re-factor.

@harshavardhana can you share details about what needs fixing? As far as I can tell, the issues mentioned in the PR itself have all been addressed, but I'm assuming minio and @dvstate had conversations out of band about what needs fixing.

It doesn't look like @dvstate is responding, but maybe someone else from the Sia community could jump in and pull this PR across the finish line. I'm interested in helping in this effort if I can get more information about what needs fixing.

**Comment by dvstate** [comment]
My apologies for not responding on here sooner. My conversations with @harshavardhana were private. All of the caching and database functionality is being stripped out. I've been tied up lately, but given a couple more days I should be able to complete all of the changes requested.

**Comment by mtlynch** [comment]
>I've been tied up lately, but given a couple more days I should be able to complete all of the changes requested.

That's awesome! I'm really excited for this integration. I think it's going to be a huge gamechanger for both Sia and minio.

**Comment by tlightsky** [comment]
@dvstate but sia need local file now, hot to deal with this if removing cache layer?(tring to use it in some case)

**Comment by ericflo** [comment]
@tlightsky I've handled this operationally for now for SiaCDN.com by dynamically creating an NFS mountable drive for each Minio instance, and mounting it on both the Sia node and the Minio node at the same mount point.

So Minio nodes each just mount their own drive (e.g. /siacache/minio1), but the Sia node has mounted:
```
/siacache/minio1
/siacache/minio2
/siacache/minio3
```

Ideally the Sia API would evolve to support buffered reader i/o endpoints in addition to filenames, but I'm not sure what the team's thinking is on that.

**Comment by tlightsky** [comment]
@ericflo not support that now, i think

**Comment by mannp** [comment]
@dvstate another 2 weeks have passed, with seemingly no github changes, so is it time to open it up for others to work on with a little more time available perhaps, or at least move things on, so others understand the code and can step in with your busy schedule.

I am thinking here of @mtlynch making the offer some time back.....

**Comment by mtlynch** [comment]
Thanks for the update, @dvstate! Excited to see this coming along.

**Comment by dvstate** [comment]
Latest commit incorporates all changes requested to date. Requesting @harshavardhana to review and provide feedback. Thank you.

**Comment by harshavardhana** [comment]
> Latest commit incorporates all changes requested to date. Requesting @harshavardhana to review and provide feedback. Thank you.

Thanks @dvstate can you rebase with latest master we can start final stage of testing and review?

File: browser/app/js/actions.js
Comment (harshavardhana) [reply browser/app/js/actions.js:496]: Can you send this change as separate PR? @dvstate do not think this belongs to this work..

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/gateway-sia.go
Comment (tlightsky) [inline cmd/gateway-sia.go]: what is this `SiaRootDir` for, i think siad should be ready before start minio

**Comment by tlightsky** [comment]
glad to see new progress, hope it be merged soon

File: cmd/gateway-main.go
Comment (tlightsky) [inline cmd/gateway-main.go]: how long will the cached files purged then?

Comment (dvstate) [reply cmd/gateway-main.go]: All cache functionality has been removed.

**Comment by dvstate** [comment]
Incorporated the latest Minio:master codebase and added Harshavardhana's changes. I also tested the latest code against Sia and the basic operations function as expected. I have sent Harshavardhana instructions for setting up Sia for testing, and will send him Siacoins to conduct the testing.

**Comment by mtlynch** [comment]
Thanks for the update, @dvstate! I tried the latest build from your branch and I can upload from the Minio browser, but not a third-party S3 client. Every file I upload fails with an error like this:

```
ERRO[2969] Unable to save an object /dvds/Last Man on Earth/Season 01/Last Man on Earth - s01e13 - Screw the Moon.mp4  cause=Sia: Object name not suitable for Sia. source=[gateway-handlers.go:314:gatewayAPIHandlers.PutObjectHandler()]
ERRO[3363] Unexpected error, please report this issue at https://github.com/minio/minio/issues  cause=Sia: Object name not suitable for Sia. source=[web-handlers.go:1100:toWebAPIError()]
```

These paths worked in 66dda37. It looks like it was URL-encoding the filenames there, whereas it's now rejecting them if they match `[^a-zA-Z0-9., _+-]+`. Would it be possible to go back to URL-encoding?

The root and cache directories are also showing up empty for me and end up dumping files in the root:

```
Sia Gateway Configuration:
  Sia Daemon API Address: 127.0.0.1:9980
  Sia Root Directory:
  Sia Cache Directory:
```

File: cmd/gateway-sia.go
Comment (mtlynch) [inline cmd/gateway-sia.go:469]: Won't `objInfo` still be uninitialized?

Comment (harshavardhana) [reply cmd/gateway-sia.go:469]: No

**Comment by dvstate** [comment]
@mtlynch Please try the latest commit to see if your third party S3 client now works with the same filename. In the past I stripped away characters that Sia couldn't recognize and just kept the original name in a local database. Now that the Sia gateway has to be stateless, I can no longer do that. If the filename is unsupported by the Sia service, then the gateway has to return an error. I haven't found an official listing of what characters Sia supports. If anyone can provide this information, I'll update the regular expression to match. For now, it's limited to what I know works.

The latest commit also changes "CacheDir" references to "TempDir" to avoid confusion. The temp directory is just used for temporary local storage. It defaults to ".sia_temp".

File: cmd/gateway-sia.go
Comment (mtlynch) [inline cmd/gateway-sia.go]: Can we add this to the `.gitignore`?

**Comment by mtlynch** [comment]
I don't have access to a Sia node at the moment, but from the code, it looks like the only restriction is leading path traversal characters (`/`, `./`, `../`).

https://github.com/NebulousLabs/Sia/blob/db53c5e60e6834b82bf529e7fd3f485cd749f51e/modules/renter/upload.go#L51

I should be able to test the latest commit tomorrow or Wednesday, but I'm fine with a conservative filename solution now that we open up after more testing.

**Comment by mtlynch** [comment]
I'm getting a build error on `make`:

```
$ make
Building minio to /home/mike/go/src/github.com/minio/minio/minio ...
# github.com/minio/minio/cmd
cmd/gateway-sia.go:435:68: undefined: HashReader
cmd/gateway-unsupported.go:44:101: undefined: HashReader
cmd/web-handlers.go:741:2: impossible type switch case: objAPI (type ObjectLayer) cannot have dynamic type *siaObjects (wrong type for PutObject method)
        have PutObject(string, string, <T>, map[string]string) (ObjectInfo, error)
        want PutObject(string, string, *"github.com/minio/minio/pkg/hash".Reader, map[string]string) (ObjectInfo, error)
Makefile:73: recipe for target 'build' failed
make: *** [build] Error 2
```

Strangely, Travis gets a different one:

```
Building minio to /home/travis/gopath/src/github.com/minio/minio/minio ...
vendor/github.com/NebulousLabs/Sia/api/api.go:9:2: cannot find package "github.com/NebulousLabs/Sia/build" in any of:
	/home/travis/gopath/src/github.com/minio/minio/vendor/github.com/NebulousLabs/Sia/build (vendor tree)
	/home/travis/.gimme/versions/go1.9.1.linux.amd64/src/github.com/NebulousLabs/Sia/build (from $GOROOT)
	/home/travis/gopath/src/github.com/NebulousLabs/Sia/build (from $GOPATH)
```

**Comment by harshavardhana** [comment]
@mtlynch I have sent a patch file to Dave to incorporate which would fix all the build issues and cleanup the history  here.  Waiting on his response.

**Comment by dvstate** [comment]
This PR has been closed as the branch commit history has been cleaned up. The PR will continue here: https://github.com/minio/minio/pull/5114

**Pull request closed without merging**
